pkg/security/certs.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// createCACertAndKey creates a CA key and a CA certificate.
+// If the certs directory does not exist, it is created.
+// If the key does not exist, it is created.
+// The certificate is written to the certs directory. If the file already exists,
+// we append the original certificates to the new certificate.
+//
+// The filename of the certificate file must be specified.
+// It should be one of:
+// - ca.crt: the general CA certificate
+// - ca-client.crt: the CA certificate to verify client certificates
+func createCACertAndKey(certsDir, caKeyPath string, caType PemUsage, keySize int, lifetime time.Duration, allowKeyReuse bool, overwrite bool) error {
+	if len(caKeyPath) == 0 {
+		return errors.New("the path to the CA key is required")
+	}
+	if len(certsDir) == 0 {
+		return errors.New("the path to the certs directory is required")
+	}
+	if caType != CAPem &&
+		caType != TenantClientCAPem &&
+		caType != ClientCAPem &&
+		caType != UICAPem {
+
+		return fmt.Errorf("caType argument to createCACertAndKey must be one of CAPem (%d), ClientCAPem (%d), or UICAPem (%d), got: %d", CAPem, ClientCAPem, UICAPem, caType)
+	}
+
+	certsDirParam := fmt.Sprintf(CERTS_DIR, certsDir)
+	caKeyParam := fmt.Sprintf(CA_KEY, caKeyPath)
+
+	// TODO do we need to move the certs to the correct path?
+	switch caType {
+	case CAPem:
+		execCmd(CREATE_CA, certsDirParam, caKeyParam)
+		//certPath = cm.CACertPath()
+	case TenantClientCAPem:
+		//certPath = cm.TenantClientCACertPath()
+	case ClientCAPem:
+		execCmd(CREATE_CLIENT, certsDirParam, caKeyParam)
+		//certPath = cm.ClientCACertPath()
+	case UICAPem:
+		//certPath = cm.UICACertPath()
+	default:
+		return errors.Newf("unknown CA type %v", caType)
+	}
+
+	//log.Infof(context.Background(), "wrote %d certificates to %s", len(certificates), certPath)
+
+	return nil
+}
+
+// CreateNodePair creates a node key and certificate.
+// The CA cert and key must load properly. If multiple certificates
+// exist in the CA cert, the first one is used.
+func CreateNodePair(certsDir, caKeyPath string, keySize int, lifetime time.Duration, overwrite bool, hosts []string) error {
+	if len(caKeyPath) == 0 {
+		return errors.New("the path to the CA key is required")
+	}
+	if len(certsDir) == 0 {
+		return errors.New("the path to the certs directory is required")
+	}
+
+	certsDirParam := fmt.Sprintf(CERTS_DIR, certsDir)
+	caKeyParam := fmt.Sprintf(CA_KEY, caKeyPath)
+	args := append(hosts, certsDirParam, caKeyParam)
+	args = append([]string{CREATE_NODE}, args...)
+	execCmd(args...)
+
+	//log.Infof(context.Background(), "generated node key: %s", keyPath)
+	return nil
+}
+
+// CreateUIPair creates a UI certificate and key using the UI CA.
+// The CA cert and key must load properly. If multiple certificates
+// exist in the CA cert, the first one is used.
+func CreateUIPair(certsDir, caKeyPath string, keySize int, lifetime time.Duration, overwrite bool, hosts []string) error {
```

Comment:
is this function to be removed? same with CreateClientPair below?

Reply:
Yes, the code should be removed, but @keith-mcclellan do we need to generate the UIPair?

Reply:
We do need to generate a UIPair for certain install patterns, particularly when we move to the 3 port config. but we can merge as is and do that change when making the port change